web/pdf_find_controller.js

```diff
@@ -86,19 +86,71 @@ const SPECIAL_CHARS_REG_EXP =
 const NOT_DIACRITIC_FROM_END_REG_EXP = /([^\p{M}])\p{M}*$/u;
 const NOT_DIACRITIC_FROM_START_REG_EXP = /^\p{M}*([^\p{M}])/u;
 
-let normalizationRegex = null;
+// The range [AC00-D7AF] corresponds to the Hangul syllables.
+// The few other chars are some CJK Compatibility Ideographs.
+const SYLLABLES_REG_EXP = /[\uAC00-\uD7AF\uFA6C\uFACF-\uFAD1\uFAD5-\uFAD7]+/g;
+const SYLLABLES_LENGTHS = new Map(
+  [
+    [...Array(0xd7af - 0xac00 + 1).keys()].map(x => x + 0xac00),
+    0xfa6c,
+    [...Array(0xfad1 - 0xfacf + 1).keys()].map(x => x + 0xfacf),
+    [...Array(0xfad7 - 0xfad5 + 1).keys()].map(x => x + 0xfad5),
+  ]
+    .flat()
+    .map(x => {
+      x = String.fromCharCode(x);
+      return [x, x.normalize("NFD").length];
+    })
+);
+// When decomposed (in using NFD) the above syllables will start
+// with one of the chars in this regexp.
+const FIRST_CHAR_SYLLABLES_REG_EXP =
+  "[\\u1100-\\u1112\\ud7a4-\\ud7af\\ud84a\\ud84c\\ud850\\ud854\\ud857\\ud85f]";
+
+let noSyllablesRegExp = null;
+let withSyllablesRegExp = null;
+
 function normalize(text) {
   // The diacritics in the text or in the query can be composed or not.
   // So we use a decomposed text using NFD (and the same for the query)
   // in order to be sure that diacritics are in the same order.
 
-  if (!normalizationRegex) {
+  // Collect syllables length and positions.
+  const syllablePositions = [];
+  let m;
+  while ((m = SYLLABLES_REG_EXP.exec(text)) !== null) {
+    let { index } = m;
+    for (const char of m[0]) {
+      syllablePositions.push([SYLLABLES_LENGTHS.get(char), index++]);
+    }
+  }
+
+  let normalizationRegex;
+  if (
+    (syllablePositions.length !== 0 || noSyllablesRegExp) &&
+    (syllablePositions.length === 0 || withSyllablesRegExp)
+  ) {
+    normalizationRegex =
+      syllablePositions.length === 0 ? noSyllablesRegExp : withSyllablesRegExp;
+  } else {
```

Comment:
This took me a moment to understand when seeing it, so how about formatting it like this instead?
```javascript
  let normalizationRegex;
  if (syllablePositions.length === 0 && noSyllablesRegExp) {
    normalizationRegex = noSyllablesRegExp;
  } else if (syllablePositions.length > 0 && withSyllablesRegExp) {
    normalizationRegex = withSyllablesRegExp;
  } else {
```